src/value.rs

```diff
@@ -158,6 +184,14 @@ impl Value {
         Value(v)
     }
 
+    /// Create a new rooted Value from an existing OCaml `value`
+    #[inline]
+    pub unsafe fn root(rt: &Runtime, v: sys::Value) -> Value {
+        let v: OCaml<'_, Value> = OCaml::new(rt, v);
+        let mut x = v.root();
+        Value::new(x.keep(v).get_raw())
```

Comment:
Be careful with this. BoxRoots are "freed" once dropped, so in this case you are not getting the behavior you are expecting. Once `x` is dropped at the end of the method, the value is not rooted anymore.

For things to work correctly the `BoxRoot` value (or another value that "owns" it) needs to be passed around.

Reply:
Thanks, I will do another pass to make sure BoxRoots are used correctly. It sounds like it would make sense for `Value` to hold a BoxRoot instead of a RawOCaml value?

Reply:
Ah but it looks like a boxroot is a pointer to an OCaml `value`, so that might not work without updating the macros (since `Value` is currently `repr(transparent)`. At that point it might just make the most sense to modify that macros to use `ocaml_interop::OCaml` directly and make `Value` an alias for `OCaml<RawValue>` (or something similar) :thinking:

Reply:
That would not work because you are in the same situation, `OCaml`, just as `Value` is just a direct representation of the raw OCaml value (but unlike `Value`, it has some extra type and lifetimes constraints attached to it), and as you noted `BoxRoot` is a pointer to that value (this ensures that you have to de-reference it for each use, which guarantees it is going to be valid, because if the GC moves the value, it will also update that pointer).

What I am thinking is that the interface you implement for `Value` could be implemented for `BoxRoot`, but there may be issues related to that which I am not considering.

Btw, here is a short recap of the various ways OCaml values are presented in `ocaml-interop`:
- `RawOCaml` just the raw value, more for internal and `unsafe` uses. Users are not meant to use this, and there are no safety guarantees. The next three are safe.
- `OCaml<'gc, T>` like `RawOCaml` but with a lifetime and type associated. The type constraint prevents the user from performing invalid operations on that value (like accessing the index of an integer). The lifetime constraint ensures that the value is not reused across calls into the OCaml runtime that may perform garbage collection. When you have a `OCaml<'gc, T>` value, you can directly manipulate the underlying OCaml value. You can also convert the value into a Rust value.
- `BoxRoot<T>` which contains a reference to an `OCaml<'gc, T>`. Needs to be dereferenced into an `OCaml<'gc, T>` to be used, and that dereferencing can only be done if the OCaml runtime handle/lock is available (this prevents code from doing anything with OCaml values when the runtime lock was released). You can convert from a `BoxRoot<T>` into a Rust value, but you cannot manipulate the value.
- `OCamlRef<'a, T>` is a reference to an `OCaml<'gc, T>` value, and it can be obtained from either an `OCaml<T>` or a `BoxRoot<T>`. This allows for a common interface that lets you pass immediate OCaml values (like fixnums, booleans, `None`, `()`, etc) without having to root them. Rust functions exported so that they are callable from `OCaml` receive these as parameters. The macro generates a wrapper that roots all the inputs so that the user doesn't have to do it. Through the runtime handle, an `OCaml<'gc, T>` can be recovered. You can convert from a `OCamlRef<'a, T>` into a Rust value, but you cannot manipulate the value.

`ocaml-rs`' `Value` as it is now can represent a not-so-safe version of `OCaml<'gc, T>`, but it cannot represent `BoxRoot<T>` or `OCamlRef<'a ,T>`. But I think the same interface you now implement for `Value` could be implemented for these (again, without the same safety guarantees you have in `ocaml-interop`, but safer than `Value` currently is).

Just rambling here, not sure yet what is the best solution here, but I think this context is relevant.

Reply:
Thank you, this is extremely helpful!

Reply:
Just added an edit with one important clarification because it was ambiguous. When I mean that "you cannot manipulate the value" in the `BoxRoot` and `OCamlRef` descriptions, I am talking about the OCaml value, not the Rust value resulting from the conversion (because there is no sharing there, data is copied).

Reply:
Btw, I think the direct equivalent of `Value` on `ocaml-interop` would be `OCaml<'static, RawValue>` (or maybe something like `Any`, using `RawValue` here to denote that it is unspecified). That would be an OCaml value with an unbounded lifetime (so, not attached to the runtime handle borrow), and without any constraints on the type of value it is.

Reply:
@zshipko thinking about this more, and one idea is to have `Value` just be a `BoxRoot<RawValue>`. You will have to handle it at the boundaries (dereferencing it in the generated code when a plain value is expected). You will not get the type-restriction safety, and will not be protected against the situation where the runtime lock is released, but will at least have something that is safe to use across possibly-gc-triggering function calls.

That is of course the approach that leaves the API being the closest to what is in there already, with some extra safety. The alternative is to go all the way and adopt `ocaml-interop`'s model, with maybe also the addition of a new `Value`-like type that doesn't discriminate over the OCaml value type (with lifetime if not implemented using `BoxRoot` or with it otherwise).

Reply:
@tizoc - I was planning on exploring the `BoxRoot` based implementation tomorrow, but the idea of tying `Value` to the runtime handle using a lifetime and adopting the `ocaml-interop` model a bit more sounds like a slightly cleaner approach!

Reply:
I had forgotten that a lot of stuff (macros, custom) depends on having the `Value` representation match `sys::Value`, so this is going to be slightly more elaborate than I expected.